[clr-interp] Fix reporting for stack slots other than the locals/arguments

davidwrighton opened the pull request:
- It turns out that the GCEncoder ignores information about tracked locals unless there is at least 1 interruptible range
- Fixing that made it clear that the current reporting strategy of using accurate reporting for those ranges wasn't correct, so instead I've enabled the conservative logic within the GC, and am now GC reporting temporary vars/call arguments via the conservative reportiung strategy.
  - Notably, the liveness analysis for local variables in the interpreter results in overlapping live ranges.
  - And arguments passed to called functions are still reporting the call arguments as well as the called function.

I believe this change works for normal functions, but I have not yet examined the actual behavior of funclets.

davidwrighton committed:
[clr-interp] Fix reporting for stack slots other than the locals/arguments
- It turns out that the GCEncoder ignores information about tracked locals unless there is at least 1 interruptible range
- Fixing that made it clear that the current reporting strategy of using accurate reporting for those ranges wasn't correct, so instead I've enabled the conservative logic within the GC, and am now GC reporting temporary vars/call arguments via the conservative reportiung strategy.
  - Notably, the liveness analysis for local variables in the interpreter results in overlapping live ranges.
  - And arguments passed to called functions are still reporting the call arguments as well as the called function.

I believe this change works for normal functions, but I have not yet examined the actual behavior of funclets.

janvorli commented:
I've hit the same issue as David last Friday. The non-tracked locals were reported in my case, but not the tracked ones.

davidwrighton commented:
@kg your initial work probably fixed a lot as most data is typically in the untracked locals, but it's not everything. However, the tracked locals work was actually not emitting anything into the actual gcinfo stream. Of course, it looks like I need to dig a little further for actual correctness as our interpreter test doesn't pass with my change. My guess is that I need to fix the funclets handling in some way. I'll look into that today once I'm in.

davidwrighton committed:
[clr-interp] Fix issue where interpreter GC stack reporting was hitting code that did comparisons against the SP pointer. We shouldn't be doing that in any situation, so use template goo to remove that from the interpreter variant of the gc info

davidwrighton committed:
Tweak range check asserts and add comment about funclet handling

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into improve_gc_reporting

davidwrighton commented:
/azp list

davidwrighton commented:
/azp run runtime-interpreter

davidwrighton committed:
Fix 3 more issues found related to crashes with invalid memory
- `InterpreterCodeManager::GetInstance` and `InterpreterCodeManager::GetParamTypeArg` needed to assume that only the SP was valid in the REGDISPLAY. Previously we used the FP to get at the values, but that isn't reliably useable during collided EH unwind first pass processing
- Tweak the behavior of `PushStackType` in the interpreter compiler so that it works correctly for pushing a value type onto the stack. This broke our implementation of mkrefany.

With these changes, we can now run the Methodical_do and Methodical_d2 test end to end.

janvorli reviewed (approved):
LGTM, thank you!

davidwrighton commented:
/ba-g The failing leg is an AzDO infrastructure issue where publish took unreasonably long, but did not fail.
